src/txn/txn_ckpt.c

```diff
@@ -767,10 +764,93 @@ __checkpoint_prepare(
 	 */
 	WT_ASSERT(session, session->ckpt_handle_next == 0);
 	WT_WITH_TABLE_READ_LOCK(session, ret = __checkpoint_apply_all(
-	    session, cfg, __wt_checkpoint_get_handles, NULL));
+	    session, cfg, __wt_checkpoint_get_handles));
 	return (ret);
 }
 
+/*
+ * __txn_checkpoint_can_skip --
+ *	Determine whether it's safe to skip taking a checkpoint. This function
+ *	also parses out some configuration options and hands them back to the
+ *	caller - make sure it does that parsing regardless of the result.
+ */
+static int
+__txn_checkpoint_can_skip(WT_SESSION_IMPL *session,
+    const char *cfg[], bool *fullp, bool *use_timestampp, bool *can_skipp)
+{
+	WT_CONFIG targetconf;
+	WT_CONFIG_ITEM cval, k, v;
+	WT_CONNECTION_IMPL *conn;
+	WT_TXN_GLOBAL *txn_global;
+	bool full, use_timestamp;
+
+	conn = S2C(session);
+	txn_global = &conn->txn_global;
+	/* Default to not skipping. */
+	*can_skipp = false;
+
+	/*
+	 * Determine if this is going to be a full checkpoint, that is a
+	 * checkpoint that applies to all data tables in a database.
+	 */
+	WT_RET(__wt_config_gets(session, cfg, "target", &cval));
+	__wt_config_subinit(session, &targetconf, &cval);
+	full = __wt_config_next(&targetconf, &k, &v) != 0;
+	if (fullp != NULL)
+		*fullp = full;
+
+	WT_RET(__wt_config_gets(session, cfg, "use_timestamp", &cval));
+	use_timestamp = cval.val != 0;
+	if (use_timestampp != NULL)
+		*use_timestampp = use_timestamp;
+
+	/* Never skip non-full checkpoints */
+	if (!full)
+		return (0);
+
+	/* Never skip if force is configured. */
+	WT_RET(__wt_config_gets_def(session, cfg, "force", 0, &cval));
+	if (cval.val != 0)
+		return (0);
+
+	/* Never skip named checkpoints. */
+	WT_RET(__wt_config_gets(session, cfg, "name", &cval));
+	if (cval.val != 0)
+		return (0);
+
+	/*
+	 * Skip checkpointing the database if nothing has been dirtied since
+	 * the last checkpoint. That said there can be short instances when a
+	 * btree gets marked dirty and the connection is yet to be. We might
+	 * skip a checkpoint in that short instance, which is okay because by
+	 * the next time we get to checkpoint, the connection would have been
+	 * marked dirty and hence the checkpoint will not be skipped this time.
+	 */
```

Comment:
"skipped this time" -> "skipped that time".
We could re-order the setting of the connection modified field I suppose, not sure why it's the way it is.